test-e2e/test_zookeeper_backup.py

```diff
@@ -0,0 +1,394 @@
+
+
+LOGGER = logging.getLogger(__name__)
+
+# Arbitrary value written to ZooKeeper.
+FLAG = b'flag'
+
+
+def _zk_set_flag(hosts: str, ephemeral: bool = False) -> str:
+    """
+    Store the `FLAG` value in ZooKeeper in a random Znode.
+    """
+    znode = '/{}'.format(uuid.uuid4())
+    zk_client = KazooClient(hosts=hosts)
+    zk_client.start()
+    zk_client.create(znode, makepath=True, ephemeral=ephemeral)
+    zk_client.set(znode, FLAG)
+    zk_client.stop()
+    return znode
+
+
+def _zk_check_flag(hosts: str, znode: str) -> None:
+    """
+    The `FLAG` value is found in ZooKeeper at the `znode` path.
+    """
+    zk_client = KazooClient(hosts=hosts, read_only=True)
+    zk_client.start()
+    value = zk_client.get(znode)
+    zk_client.stop()
+    assert value[0] == FLAG
+
+
+def _bash(cmd: List[str]) -> None:
+    """
+    Execute `cmd` in a subprocess, log potential errors.
+    """
+    LOGGER.debug('BASH ARGS: {}'.format(' '.join(cmd)))
+    try:
+        result = subprocess.check_output(
+            args=' '.join(cmd),
+            stderr=subprocess.STDOUT,
+            shell=True,
+        )
+        LOGGER.debug(result.decode())
+    except subprocess.CalledProcessError as exc:
+        LOGGER.error(exc.output.decode())
+        raise
+
+
+BACKUP_SCRIPT = textwrap.dedent(
+    """\
+#!/bin/bash -e
+
+DESTINATION_DIR=$PWD
+REMOTE_TMP_DIR=/tmp/zk_backup
+
+usage() {
+  echo "Usage: $0 [-h] [-p ZK_BACKUP_DEST_DIR] [-i SSH_KEY_PATH] [-t REMOTE_TMP_DIR] SSH_USER@MASTER_HOST" 1>&2
+}
+exit_abnormal() {
+  usage
+  exit 1
+}
+while getopts ":hp:t:i:" opt; do
+  case $opt in
+    h)
+      usage
+      exit 0
+    ;;
+    p) DESTINATION_DIR=$OPTARG
+    ;;
+    t) REMOTE_TMP_DIR=$OPTARG
+    ;;
+    i) SSH_KEY_PATH=$OPTARG
+    ;;
+    \\?) echo "Invalid option -$OPTARG" 1>&2
+        exit_abnormal
+    ;;
+    :) echo "Error: -$OPTARG requires an argument." 1>&2
+       exit_abnormal
+    ;;
+    *) exit_abnormal
+    ;;
+  esac
+done
+
+shift $((OPTIND - 1))
+if (($# != 1)); then
+  exit_abnormal
+fi
+
+CONN=$1
+KEY=${SSH_KEY_PATH:+"-i $SSH_KEY_PATH"}
+OPTS='-o ConnectTimeout=3 -o ConnectionAttempts=1
+-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no'
+
+ssh $OPTS $KEY -tt $CONN << EOF
+  set -e
+  sudo mkdir -p $REMOTE_TMP_DIR
+  sudo systemctl stop dcos-exhibitor
+  sudo cp -pr /var/lib/dcos/exhibitor/zookeeper ${REMOTE_TMP_DIR}/zookeeper
+  sudo systemctl start dcos-exhibitor
+  sudo tar --exclude 'myid' --exclude 'zookeeper.out' -pcvzf \\
+  ${REMOTE_TMP_DIR}/zk_backup-'$(date +%Y-%m-%d_%H-%M-%S)'.tar.gz -C $REMOTE_TMP_DIR ./zookeeper
+  sudo rm -rf ${REMOTE_TMP_DIR}/zookeeper
+  exit 0
+EOF
+
+scp $OPTS $KEY -r ${CONN}:${REMOTE_TMP_DIR}/zk_backup-'*'.tar.gz ${DESTINATION_DIR}/ 2> /dev/null
+
+ssh $OPTS $KEY -tt $CONN << EOF
+  set -e
+  sudo rm -rf $REMOTE_TMP_DIR
+  exit 0
+EOF
+exit 0
+    """
+)
+
+
+RESTORE_SCRIPT = textwrap.dedent(
+    """\
+#!/bin/bash -e
+
+SSH_USER='root'
```

Comment:
Why are the user@host ssh details configured differently to backup? Can we make them more symmetric?

Reply:
:+1: